lib/src/search/search_navigation_controller.dart

```diff
@@ -21,75 +21,88 @@ import 'widget/search_navigation_widget.dart';
 /// enters the editing mode where it doesn't advance the [value] to currentMatch
 /// nor change the selection of the [CodeController].
 ///
-/// Also used to manage the state of [SearchNavigationWidget]
+/// Also used to manage the state of [SearchNavigationWidget].
 class SearchNavigationController extends ValueNotifier<SearchNavigationState> {
   final CodeController codeController;
 
   SearchResult _searchResult = SearchResult.empty;
   String _lastText = '';
-  bool _isEditing = false;
+  bool _wasEdited = false;
 
   SearchNavigationController({
     required this.codeController,
     SearchNavigationState? state,
-  }) : super(state ?? SearchNavigationState()) {
+  }) : super(state ?? SearchNavigationState.noMatches) {
     codeController.addListener(_updateState);
     _lastText = codeController.code.text;
   }
 
   void moveNext() {
-    _isEditing = false;
+    _wasEdited = false;
     if (_searchResult.matches.isEmpty) {
       return;
     }
 
-    if (value.currentMatchIndex == _searchResult.matches.length - 1) {
-      value = value.copyWith(currentMatchIndex: 0);
-    } else {
-      value = value.copyWith(currentMatchIndex: value.currentMatchIndex + 1);
-    }
+    final currentIndex =
+        value.currentMatchIndex ?? _getClosestMatchIndex() ?? 0;
+
+    value = value.copyWith(
+      currentMatchIndex: (currentIndex + 1) % value.totalMatchesCount,
+    );
 
-    moveSelectionToMatch(value.currentMatchIndex);
+    _moveSelectionToMatch(value.currentMatchIndex!);
   }
 
   void movePrevious() {
-    _isEditing = false;
+    _wasEdited = false;
     if (_searchResult.matches.isEmpty) {
       return;
     }
 
-    if (value.currentMatchIndex == 0) {
-      value = value.copyWith(
-        currentMatchIndex: _searchResult.matches.length - 1,
-      );
-    } else {
-      value = value.copyWith(currentMatchIndex: value.currentMatchIndex - 1);
-    }
+    final currentIndex =
+        value.currentMatchIndex ?? _getClosestMatchIndex() ?? 0;
 
-    moveSelectionToMatch(value.currentMatchIndex);
+    value = value.copyWith(
+      currentMatchIndex: (currentIndex - 1) % value.totalMatchesCount,
+    );
+
+    _moveSelectionToMatch(value.currentMatchIndex!);
   }
 
   void _updateState() {
+    if (codeController.fullSearchResult == _searchResult) {
+      return;
+    }
+
+    _searchResult = codeController.fullSearchResult;
+
+    value = value.copyWith(totalMatchesCount: _searchResult.matches.length);
```

Comment:
If we set `this.value` multiple times, it causes that many `notifyListeners` calls.
To do this in one step, make `value = _createValue();`